### PR TITLE
contribution guide

### PR DESCRIPTION
**Pull request by reactormonk**
Fixes e.g. https://github.com/nim-lang/Nim/issues/1606 and should help new contributors to get started. Style guides etc. Could be added too. I still haven't figured out how to compare tests from your branch to devel 100%, so feel free to add that if you know how.

The branch is on the nim-lang repo, so feel free to push to it / create a PR.

### CODE REVIEW

File: contributing.md [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+Not all the tests follow this scheme, feel free to change the ones
+that don't. Always leave the code cleaner than you found it.
+
+## Stdlib
+
+If you change the stdlib (anything under `lib/`), put a test in the
+file you changed. Add the tests under an `when isMainModule:`
+condition so they only get executed when the tester is building the
+file. Each test should be in a separate `block:` statement, such that
+each has its own scope. Use boolean conditions and `assert` for the
```
Comment (Araq) [inline contributing.md]: doAssert should be used instead of assert.

File: contributing.md [hunk 2]
```diff
@@ -0,0 +1,101 @@
+# The git stuff
```
Comment (Araq) [inline contributing.md]: Meh, usually we use reStructuredText instead of Markdown... Markdown is trailing whitespace sensitive and thus evil.

Comment (reactormonk) [inline contributing.md]: Sure, but github picks up `contributing.md`. Windmills etc.

Comment (kirbyfan64) [inline contributing.md]: @reactormonk https://www.google.com/search?q=github+contributing.rst&oq=github+contributing.rst

**Comment by yglukhov** [comment]
How about starting using unittests.nim for unittesting? ;)

File: contributing.md (hunk 2 above)
Comment (reactormonk) [inline contributing.md]: @kirbyfan64 neat, could you make a PR against this branch converting it to rst?

**Comment by reactormonk** [comment]
@yglukhov Now you know where to document it once you've reached a stable enough state.

**Comment by reactormonk** [comment]
The hyperlinks don't seem to work as expected - what's the correct syntax?

**Comment by kirbyfan64** [comment]
@reactormonk Try putting a space between the link text and the left arrow (`<`).

File: contributing.rst [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+- exitcode: Exit code of the test (via ``exit(number)``)
+- errormsg: The expected error message
+- file: The file the errormsg
+- line: The line the errormsg was produced at
+
+An example for a test:
+
+.. code-block:: nim
+
+  discard """
```
Comment (dom96) [inline contributing.rst]: Where does this ``discard`` end?

**Comment by dom96** [comment]
Would be nice to mention that you can run certain categories of tests only with the tester.

Also, the ``tests/readme.md`` file should be improved to contain the information about the keys, or at least a link to this document. But that can be a separate PR.

**Comment by reactormonk** [comment]
@dom96 but how to know which categories to be tested? I don't think that relevant for your average contributor.

**Comment by dom96** [comment]
@reactormonk It's a useful bit of knowledge. Just mention it briefly.

**Comment by reactormonk** [comment]
Without an explanation what the categories are it's not really useful. I don't think a basic guide should cover it. As soon as you start messing with things that have its own test category you'll find it.

**Comment by dom96** [comment]
You don't need an explanation of the categories, you will see them when you open the directory anyway. Why shouldn't it be covered?

File: contributing.rst (hunk 3 above)
Comment (Araq) [inline contributing.rst]: Indeed, close the discard properly.

**Pull request merged**